tests/integration/test_groupby.py

```diff
@@ -405,17 +405,26 @@ def test_groupby_split_out(c, input_table, split_out, request):
     assert return_df.npartitions == split_out if split_out else 1
     assert_eq(return_df.sort_values("user_id"), expected_df, check_index=False)
 
+    return_df = c.sql(
+        f"""
+        SELECT DISTINCT(user_id) FROM {input_table}
+        """,
+        config_options={"sql.groupby.split_out": split_out},
+    )
+    expected_df = user_table[["user_id"]].drop_duplicates()
+    assert return_df.npartitions == split_out if split_out else 1
+    assert_eq(return_df.sort_values("user_id"), expected_df, check_index=False)
+
 
-@pytest.mark.skip(reason="WIP DataFusion")
 @pytest.mark.parametrize(
     "gpu,split_every,expected_keys",
     [
-        (False, 2, 74),
-        (False, 3, 68),
-        (False, 4, 64),
-        pytest.param(True, 2, 107, marks=pytest.mark.gpu),
-        pytest.param(True, 3, 101, marks=pytest.mark.gpu),
-        pytest.param(True, 4, 97, marks=pytest.mark.gpu),
+        (False, 2, 72),
+        (False, 3, 66),
+        (False, 4, 62),
+        pytest.param(True, 2, 105, marks=pytest.mark.gpu),
+        pytest.param(True, 3, 99, marks=pytest.mark.gpu),
+        pytest.param(True, 4, 95, marks=pytest.mark.gpu),
```

Comment:
Have probably asked about this before, but is there any way we could modify this test so that we don't have to modify this param when the aggregate codepath is changed?

Perhaps instead of checking for an exact number of keys, we could assert that we don't have a number of keys that would indicate the `split_every` param is being ignored?

Reply:
I've updated the tests here to specifically check for the reduction tasks in the dask graph instead of the overall task graph. The expected number of keys can now be computed based on a utility function.

Reply:
> The expected number of keys can now be computed based on a utility function.

Wouldn't this offload the issue of needing to change the params to the utility function?

I guess what I am looking for here is tests that ideally wouldn't need to be modified at all if we changed something in the groupby agg codepath, which still seems to be an issue with https://github.com/dask-contrib/dask-sql/pull/575/commits/11b0f50ac17719b75ed61f9d279958fdeea35b84 - removing the groupby aggs from the `drop_duplicates` call causes tests to fail.

If checking for exact number of keys is needed here, then maybe it makes more sense to keep the explicit parameters, as that's a little easier to update for users making changes to the groupby codepath that might not understand the underlying logic behind the number of tree reduce operations. Otherwise, maybe we could do something like grabbing the relevant groupby layers from the HLG and checking that they have the anticipated `split_out`/`split_every` set?

Reply:
I looked into this a bit more and here's what I found:

> Wouldn't this offload the issue of needing to change the params to the utility function?

The utility function just applies some math to compute the theoretical number of operations in a tree reduction step given number of partitions and `split_every`. This number should theoretically be independent of Dask-sql's codepath and should remain unchanged provided Dask's approach to groupby doesn't change significantly. Though it doesn't handle the `None` case very well. 

> Otherwise, maybe we could do something like grabbing the relevant groupby layers from the HLG and checking that they have the anticipated split_out/split_every set?

The hlg layer if doing this on Dask DataFrame does contain information about `split_every` that we can check directly but since the codepath for dask-cudf groupby's are slightly different this information is not available for the gpu case. 

One option as discussed offline could be to remove `split_every` as a parameter and instead run the query multiple times with different values within the pytest asserting that `num_keys(split_every=2)>=num_keys(split_every=3)..` and so on.

Do you have a preference of this vs keeping the current approach as is. 

> removing the groupby aggs from the drop_duplicates call causes tests to fail.

Could you elaborate a bit more on this point.

Reply:
> Could you elaborate a bit more on this point

Apologies for the vague wording 🙂 my initial reason around starting this thread was because the changes made in https://github.com/dask-contrib/dask-sql/pull/575/commits/37d61d5317d11d36f7e9889d88ad6822ee645678 caused the number of Dask keys in this test to change, breaking the test and forcing us to modify the parameters; when I suggested modifying the test, it was with the hope that we could make a `split_every` test that passes independently of changes to our groupby agg codepath, such that with or without that commit it would pass.

> One option as discussed offline could be to remove split_every as a parameter and instead run the query multiple times with different values within the pytest asserting that num_keys(split_every=2)>=num_keys(split_every=3)..

I think something like this could work here if we don't have a larger reason to parametrize on `split_every` 🙂

Reply:
Thanks for clarifying. I think moving away from parametrizing on split_every might be more robust in this case. 

Just wanted to clarify that the changes made in this pr don't actually modify the number of keys. The new datafusion branch has a slightly different task graph vs the calcite one which results in different number of keys. 
Though this does expose a flaw in these tests where anytime our task graph changes (not only related to groupby) the number of keys would need to be updated).